File: app/src/main/java/br/com/example/meuprimeiroexemplo/resource/PostResource.java
package br.com.example.meuprimeiroexemplo.resource;

import java.util.List;

import br.com.example.meuprimeiroexemplo.model.Post;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.HEAD;
import retrofit2.http.OPTIONS;
import retrofit2.http.PATCH;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;

public interface PostResource<T, E> {

    @POST("posts")
    Call<Post> post(@Body Post post);

    @GET("{id}")
    Call<Post> get(@Path("id") Integer id);

    @GET("posts")
    Call<List<Post>> get();

    @PUT("{id}")
    Call<Void> put(@Path("id") Integer id, @Body Post post);

    @PATCH("{id}")
    Call<Void> patch(@Path("id") Integer id);

    @DELETE("{id}")
    Call<Void> delete(@Path("id") Integer id);

    @OPTIONS
    Call<Void> options();

    @HEAD
    Call<Void> head();
}

File: app/src/main/java/br/com/example/meuprimeiroexemplo/resource/PostResource.kt
package br.com.example.meuprimeiroexemplo.resource

import br.com.example.meuprimeiroexemplo.model.Post
import retrofit2.Call
import retrofit2.http.*

interface PostResource {
    @POST("posts")
    fun post(@Body post: Post?): Call<Post?>?

    @GET("posts/{id}")
    operator fun get(@Path("id") id: Int?): Call<Post?>?

    @GET("posts")
    fun get(): Call<List<Post?>?>?

    @PUT("posts/{id}")
    fun put(@Path("id") id: Int, @Body post: Post)

    @PATCH("posts/{id}")
    fun patc(@Path("id") id: Int?, @Body post: Post?): Call<Void?>?

    @DELETE("posts/{id}")
    fun delete(@Path("id") id: Int?): Call<Void?>?
}